# Déjame escuchar la música

# Contenido <a id='back'></a>

* [Introducción](#intro)
* [Etapa 1. Descripción de los datos](#data_review)
    * [Conclusiones](#data_review_conclusions)
* [Etapa 2. Preprocesamiento de datos](#data_preprocessing)
    * [2.1 Estilo del encabezado](#header_style)
    * [2.2 Valores ausentes](#missing_values)
    * [2.3 Duplicados](#duplicates)
    * [2.4 Conclusiones](#data_preprocessing_conclusions)
* [Etapa 3. Prueba de hipótesis](#hypothesis)
    * [3.1 Hipótesis 1: actividad de los usuarios y las usuarias en las dos ciudades](#activity)
* [Conclusiones](#end)

## Introducción <a id='intro'></a>
---

Como analista de datos, tu trabajo consiste en analizar datos para extraer información valiosa y tomar decisiones basadas en ellos. Esto implica diferentes etapas, como la descripción general de los datos, el preprocesamiento y la prueba de hipótesis.

Siempre que investigamos, necesitamos formular hipótesis que después podamos probar. A veces aceptamos estas hipótesis; otras veces, las rechazamos. Para tomar las decisiones correctas, una empresa debe ser capaz de entender si está haciendo las suposiciones correctas.

En este proyecto, compararás las preferencias musicales de las ciudades de Springfield y Shelbyville. Estudiarás datos reales de transmisión de música online para probar la hipótesis a continuación y comparar el comportamiento de los usuarios y las usuarias de estas dos ciudades.

### Objetivo:
Prueba la hipótesis:
1. La actividad de los usuarios y las usuarias difiere según el día de la semana y dependiendo de la ciudad.


### Etapas
Los datos del comportamiento del usuario se almacenan en el archivo `/datasets/music_project_en.csv`. No hay ninguna información sobre la calidad de los datos, así que necesitarás examinarlos antes de probar la hipótesis.

Primero, evaluarás la calidad de los datos y verás si los problemas son significativos. Entonces, durante el preprocesamiento de datos, tomarás en cuenta los problemas más críticos.

Tu proyecto consistirá en tres etapas:
 1. Descripción de los datos.
 2. Preprocesamiento de datos.
 3. Prueba de hipótesis.








[Volver a Contenidos](#back)

## Etapa 1. Descripción de los datos <a id='data_review'></a>
---


### Exploración de los datos  

1. Abre los datos y examínalos. Necesitarás `pandas`, así que impórtalo.

In [4]:
# Importar pandas
import pandas as pd

2. Lee el archivo `music_project_en.csv` de la carpeta `/datasets/` y guárdalo en la variable `df`:

In [5]:
# Leer el archivo y almacenarlo en df
df = pd.read_csv("../data/music_project_en.csv")

3. Muestra las 10 primeras filas de la tabla:

In [6]:
# Obtener las 10 primeras filas de la tabla df
print(df.head(10))

     userID                        Track            artist   genre  \
0  FFB692EC            Kamigata To Boots  The Mass Missile    rock   
1  55204538  Delayed Because of Accident  Andreas Rönnberg    rock   
2    20EC38            Funiculì funiculà       Mario Lanza     pop   
3  A3DD03C9        Dragons in the Sunset        Fire + Ice    folk   
4  E2DC1FAE                  Soul People        Space Echo   dance   
5  842029A1                       Chains          Obladaet  rusrap   
6  4CB90AA5                         True      Roman Messer   dance   
7  F03E1C1F             Feeling This Way   Polina Griffith   dance   
8  8FA1D3BE                     L’estate       Julia Dalia  ruspop   
9  E772D5C0                    Pessimist               NaN   dance   

        City        time        Day  
0  Shelbyville  20:28:33  Wednesday  
1  Springfield  14:07:09     Friday  
2  Shelbyville  20:58:07  Wednesday  
3  Shelbyville  08:37:09     Monday  
4  Springfield  08:34:34     Monday  
5

4. Obtén la información general sobre la tabla con un comando. Conoces el método que muestra la información general que necesitamos.

In [7]:
# Obtener la información general sobre nuestros datos
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 65079 entries, 0 to 65078
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0     userID  65079 non-null  object
 1   Track     63736 non-null  object
 2   artist    57512 non-null  object
 3   genre     63881 non-null  object
 4     City    65079 non-null  object
 5   time      65079 non-null  object
 6   Day       65079 non-null  object
dtypes: object(7)
memory usage: 3.5+ MB


In [8]:
# Revisar el nombre de las columnas 
print(df.columns)

Index(['  userID', 'Track', 'artist', 'genre', '  City  ', 'time', 'Day'], dtype='object')


### Observaciones generales

Estas son nuestras observaciones sobre la tabla. Contiene siete columnas. Almacenan los mismos tipos de datos: `object`.

Según la documentación:
- `' userID'`: identificador del usuario o la usuaria;
- `'Track'`: título de la canción;
- `'artist'`: nombre del artista;
- `'genre'`: género de la pista;
- `'City'`: ciudad del usuario o la usuaria;
- `'time'`: la hora exacta en la que se reprodujo la canción;
- `'Day'`: día de la semana.

Podemos ver tres problemas con el estilo en los encabezados de la tabla:
1. Algunos encabezados están en mayúsculas, otros en minúsculas.
2. Hay espacios en algunos encabezados.
3. Parece que el formato de `userID` es camelCase, siendo distinto a los demás. 
4. Existe un espacio al final de la columna `City`.
5. **Opcional**. En lo personal, la columna `time` es muy genérica, la nombraría como `track_play_time` para tener un mejor entendimiento de lo que la variable describe. 


### Escribe observaciones de tu parte. Estas son algunas de las preguntas que pueden ser útiles: <a id='data_review_conclusions'></a>

1. ¿Qué tipo de datos tenemos a nuestra disposición en las filas? 

In [9]:
## Conocer el tipo de datos que se almacenan en el df
print(df.dtypes)

  userID    object
Track       object
artist      object
genre       object
  City      object
time        object
Day         object
dtype: object


¿Y cómo podemos entender lo que almacenan las columnas?

In [10]:
## Conocer lo que almacenan las columnas
print(df.describe())

          userID  Track     artist  genre       City        time     Day
count      65079  63736      57512  63881        65079     65079   65079
unique     41748  39666      37806    268            2     20392       3
top     A8AE9169  Brand  Kartvelli    pop  Springfield  08:14:07  Friday
freq          76    136        136   8850        45360        14   23149


2. ¿Hay suficientes datos para proporcionar respuestas a nuestra hipótesis o necesitamos más información?

In [11]:
## Conocer la dimensión de la df 
print(df.shape)

(65079, 7)


3. ¿Notaste algún problema en los datos, como valores ausentes, duplicados o tipos de datos incorrectos?

In [12]:
## Valores ausentes

### Número de valores ausentes por columna
print(df.isna().sum())

### Porcentaje de valores ausentes por columna
print(df.isna().mean()*100)

  userID       0
Track       1343
artist      7567
genre       1198
  City         0
time           0
Day            0
dtype: int64
  userID     0.000000
Track        2.063646
artist      11.627407
genre        1.840840
  City       0.000000
time         0.000000
Day          0.000000
dtype: float64


In [13]:
## Valores duplicados explícitos

## Número de datos duplicados
print(df.duplicated().sum())
print()

## Identificar filas duplicadas
print(df[df.duplicated()])

3826

         userID                Track                          artist  \
575    E7F07B46                Crazy                  The Manhattans   
832    7671A47A           Johnny Go!                          NikitA   
1216   69467B01        Change It All                  Harrison Storm   
1754   13B1A573    Te Adoramos Jesús                         Athenas   
1964   B24668A0  Mad over You Mashup                      Nana Fofie   
...         ...                  ...                             ...   
65042  83E9C8C4        Buddhist Beat  Asian Zen Spa Music Meditation   
65056  2E25BCD2          Psychobitch                          Easter   
65059  F231C47E  All Summer in a Day                     VHS Or BETA   
65067  F1B93F29          Poison Kiss                     Centerstone   
65069  BE1AAD74            Waterwalk                Eduardo Gonzales   

            genre       City        time        Day  
575           rnb  Springfield  13:39:46     Monday  
832        ruspop  Sp

Creamos una nueva dataframe eliminando los duplicados explícitos y la volvemos a analizar. 

In [14]:
## Eliminar duplicados explícitos
df_clean = df.drop_duplicates().reset_index(drop= True)

In [15]:
## Conocer composición
print(df_clean.describe())

          userID  Track     artist  genre       City        time     Day
count      61253  59991      54156  60126        61253     61253   61253
unique     41748  39666      37806    268            2     20392       3
top     A8AE9169  Brand  Kartvelli    pop  Springfield  08:14:07  Friday
freq          71    127        127   8323        42741        13   21840


### Conclusión 

El análisis del DataFrame revela un total de 65,079 observaciones. Inicialmente, todas las columnas parecen estar catalogadas como tipo `object`. Esto podría deberse a varios factores, como una importación incorrecta de datos, datos mixtos en las columnas o una interpretación inadecuada de los datos. Sin embargo, tras revisar las primeras 10 observaciones, hemos descartado problemas relacionados con la importación de datos.

De las siete variables presentes en el DataFrame, tres presentan valores faltantes. La columna `artist` tiene el mayor porcentaje de datos faltantes (11.63%), seguida por `Track` (2.06%) y `genre` (1.84%).

En cuanto a los valores duplicados, se identificaron un total de 3,826 filas duplicadas, lo que representa aproximadamente el 5.88% del total del DataFrame. Para investigar posibles duplicados intrínsecos, creamos un nuevo DataFrame eliminando los duplicados extrínsecos y realizamos un análisis adicional. Observamos una sobrerepresentación de ciertos valores, pero esto no necesariamente indica duplicados implícitos. Se requiere un análisis más detallado para cada columna.

Dado que los datos faltantes y duplicados representan una pequeña fracción del DataFrame, consideramos que contamos con una muestra poblacional adecuada para llevar a cabo nuestros análisis.

[Volver a Contenidos](#back)

## Etapa 2. Preprocesamiento de datos <a id='data_preprocessing'></a>
---

El objetivo aquí es preparar los datos para que sean analizados.
El primer paso es resolver cualquier problema con los encabezados. Luego podemos avanzar a los valores ausentes y duplicados. Empecemos.

Corrige el formato en los encabezados de la tabla.


### Estilo del encabezado <a id='header_style'></a>
Muestra los encabezados de la tabla (los nombres de las columnas):

In [16]:
# Muestra los nombres de las columnas
print(df.columns)

Index(['  userID', 'Track', 'artist', 'genre', '  City  ', 'time', 'Day'], dtype='object')


Cambia los encabezados de la tabla de acuerdo con las reglas del buen estilo:
* Todos los caracteres deben ser minúsculas.
* Elimina los espacios.
* Si el nombre tiene varias palabras, utiliza snake_case.

Anteriormente, aprendiste acerca de la forma automática de cambiar el nombre de las columnas. Vamos a aplicarla ahora. Utiliza el bucle for para iterar sobre los nombres de las columnas y poner todos los caracteres en minúsculas. Cuando hayas terminado, vuelve a mostrar los encabezados de la tabla:

In [17]:
# Bucle en los encabezados poniendo todo en minúsculas
new_col_names = []

for old_name in df.columns: 
    name_lowered = old_name.lower()
    new_col_names.append(name_lowered)
    
df.columns = new_col_names 
print(df.columns)


Index(['  userid', 'track', 'artist', 'genre', '  city  ', 'time', 'day'], dtype='object')


Ahora, utilizando el mismo método, elimina los espacios al principio y al final de los nombres de las columnas e imprime los nombres de las columnas nuevamente:

In [18]:
# Bucle en los encabezados eliminando los espacios
new_col_names = []

for old_name in df.columns: 
    name_stripped = old_name.strip()
    new_col_names.append(name_stripped)
    
df.columns = new_col_names 
print(df.columns)

Index(['userid', 'track', 'artist', 'genre', 'city', 'time', 'day'], dtype='object')


Necesitamos aplicar la regla de snake_case a la columna `userid`. Debe ser `user_id`. Cambia el nombre de esta columna y muestra los nombres de todas las columnas cuando hayas terminado.

In [19]:
# Cambiar el nombre de la columna "userid"
df.rename(columns = {"userid": "user_id"}, inplace= True)


Comprueba el resultado. Muestra los encabezados una vez más:

In [20]:
# Comprobar el resultado: la lista de encabezados
print(df.columns)

Index(['user_id', 'track', 'artist', 'genre', 'city', 'time', 'day'], dtype='object')


[Volver a Contenidos](#back)

### Valores ausentes <a id='missing_values'></a>
 Primero, encuentra el número de valores ausentes en la tabla. Debes utilizar dos métodos en una secuencia para obtener el número de valores ausentes.

In [21]:
# Calcular el número de valores ausentes

## Valores ausentes por columa
missing_by_column = df.isna().sum()
print(missing_by_column)
print() # Espacio

## Valores ausentes totales
print("Datos faltantes totales: ", missing_by_column.sum())

user_id       0
track      1343
artist     7567
genre      1198
city          0
time          0
day           0
dtype: int64

Datos faltantes totales:  10108


No todos los valores ausentes afectan a la investigación. Por ejemplo, los valores ausentes en `track` y `artist` no son cruciales. Simplemente puedes reemplazarlos con valores predeterminados como el string `'unknown'` (desconocido).

Pero los valores ausentes en `'genre'` pueden afectar la comparación entre las preferencias musicales de Springfield y Shelbyville. En la vida real, sería útil saber las razones por las cuales hay datos ausentes e intentar recuperarlos. Pero no tenemos esa oportunidad en este proyecto. Así que tendrás que:
* rellenar estos valores ausentes con un valor predeterminado;
* evaluar cuánto podrían afectar los valores ausentes a tus cómputos;

Reemplazar los valores ausentes en las columnas `'track'`, `'artist'` y `'genre'` con el string `'unknown'`. Como mostramos anteriormente en las lecciones, la mejor forma de hacerlo es crear una lista que almacene los nombres de las columnas donde se necesita el reemplazo. Luego, utiliza esta lista e itera sobre las columnas donde se necesita el reemplazo haciendo el propio reemplazo.

In [22]:
# Bucle en los encabezados reemplazando los valores ausentes con 'unknown'
## Definir las columnas a inspeccionar
columns_to_replace = ["track", "artist", "genre"]

## Bucle para remplazar 
for col in columns_to_replace: 
    df[col].fillna("unknown", inplace=True)

/var/folders/_8/p4kd7db956n4lq8gs8sy4xv80000gp/T/ipykernel_71995/3540213823.py:7: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df[col].fillna("unknown", inplace=True)


Ahora comprueba el resultado para asegurarte de que después del reemplazo no haya valores ausentes en el conjunto de datos. Para hacer esto, cuenta los valores ausentes nuevamente.

In [23]:
# Contar valores ausentes
## Por columna
missing_by_column = df.isna().sum()
print(missing_by_column)
print() # Espacio

## Totales
print("Datos faltantes totales: ", missing_by_column.sum())

user_id    0
track      0
artist     0
genre      0
city       0
time       0
day        0
dtype: int64

Datos faltantes totales:  0


[Volver a Contenidos](#back)

### Duplicados <a id='duplicates'></a>
Encuentra el número de duplicados explícitos en la tabla. Una vez más, debes aplicar dos métodos en una secuencia para obtener la cantidad de duplicados explícitos.

In [24]:
# Contar duplicados explícitos
## 1. Identificar los datos duplicados
duplicated_data = df.duplicated()

print(duplicated_data.sum())

3826


Ahora, elimina todos los duplicados. Para ello, llama al método que hace exactamente esto.

In [25]:
# Eliminar duplicados explícitos
df.drop_duplicates(inplace= True)

Comprobemos ahora si eliminamos con éxito todos los duplicados. Cuenta los duplicados explícitos una vez más para asegurarte de haberlos eliminado todos:

In [26]:
# Comprobar de nuevo si hay duplicados
print(df.duplicated().sum())

0


Ahora queremos deshacernos de los duplicados implícitos en la columna `genre`. Por ejemplo, el nombre de un género se puede escribir de varias formas. Dichos errores también pueden afectar al resultado.

Para hacerlo, primero mostremos una lista de nombres de género únicos, ordenados en orden alfabético. Para ello:
* Extrae la columna `genre` del DataFrame.
* Llama al método que devolverá todos los valores únicos en la columna extraída.


In [27]:
# Inspeccionar los nombres de géneros únicos
## Extracción columna `genre`
genre_colunm= df["genre"]

## Número de géneros únicos
print(genre_colunm.nunique())

## Alternativa en un solo comando
print(df["genre"].nunique())

269
269


Busca en la lista para encontrar duplicados implícitos del género `hiphop`. Estos pueden ser nombres escritos incorrectamente o nombres alternativos para el mismo género.

In [28]:
## Conocer el número de veces que aparece "hiphop"
print(df[df["genre"] == "hiphop"].shape[0])

79


In [29]:
## Conocer el número de veces que aparece "hip" "hop" "hip-hop"
hiphop_misspelled = ["hip", "hop", "hip-hop"]

for name in hiphop_misspelled: 
    count = df[df["genre"] == name].shape[0]
    print(f'"{name}" appeared {count} times')
    

"hip" appeared 2975 times
"hop" appeared 1 times
"hip-hop" appeared 1 times


Verás los siguientes duplicados implícitos:
* `hip`
* `hop`
* `hip-hop`

Para deshacerte de ellos, crea una función llamada `replace_wrong_genres()` con dos parámetros:
* `wrong_genres=`: esta es una lista que contiene todos los valores que necesitas reemplazar.
* `correct_genre=`: este es un string que vas a utilizar como reemplazo.

Como resultado, la función debería corregir los nombres en la columna `'genre'` de la tabla `df`, es decir, remplazar cada valor de la lista `wrong_genres` por el valor en `correct_genre`.

Dentro del cuerpo de la función, utiliza un bucle `'for'` para iterar sobre la lista de géneros incorrectos, extrae la columna `'genre'` y aplica el método `replace` para hacer correcciones.

In [30]:
# Función para reemplazar duplicados implícitos
def replace_wrong_genres(wrong_genres, correct_genre): 
        
       for wrong_genre in wrong_genres: 
            df["genre"] = df["genre"].replace(wrong_genre, correct_genre)
       return df

Ahora, llama a `replace_wrong_genres()` y pásale tales argumentos para que retire los duplicados implícitos (`hip`, `hop` y `hip-hop`) y los reemplace por `hiphop`:

In [31]:
# Eliminar duplicados implícitos
## Definir variables
misspelled = ["hip", "hop", "hip-hop"]
correct_name = "hiphop"

## Asignación dataframe nuevo
df_corrected = replace_wrong_genres(wrong_genres = misspelled, correct_genre= correct_name)

Asegúrate de que los nombres duplicados han sido eliminados. Muestra la lista de valores únicos de la columna `'genre'` una vez más:

In [32]:
## Nombres duplicados 
print(df_corrected["genre"].nunique())

266


In [33]:
# Comprobación de duplicados implícitos
print(df_corrected[df_corrected["genre"] == "hiphop"].shape[0])

3056


[Volver a Contenidos](#back)

### Tus observaciones <a id='data_preprocessing_conclusions'></a>

`Describe brevemente lo que has notado al analizar duplicados, cómo abordaste sus eliminaciones y qué resultados obtuviste.`
Estoy aqui 

[Volver a Contenidos](#back)

## Etapa 3. Prueba de hipótesis <a id='hypothesis'></a>

### Hipótesis: comparar el comportamiento del usuario o la usuaria en las dos ciudades <a id='activity'></a>

La hipótesis afirma que existen diferencias en la forma en que los usuarios y las usuarias de Springfield y Shelbyville consumen música. Para comprobar esto, usa los datos de tres días de la semana: lunes, miércoles y viernes.

* Agrupa a los usuarios y las usuarias por ciudad.
* Compara el número de canciones que cada grupo reprodujo el lunes, el miércoles y el viernes.


Realiza cada cálculo por separado.

El primer paso es evaluar la actividad del usuario en cada ciudad. Recuerda las etapas dividir-aplicar-combinar de las que hablamos anteriormente en la lección. Tu objetivo ahora es agrupar los datos por ciudad, aplicar el método apropiado para contar durante la etapa de aplicación y luego encontrar la cantidad de canciones reproducidas en cada grupo especificando la columna para obtener el recuento.

A continuación se muestra un ejemplo de cómo debería verse el resultado final:
`df.groupby(by='....')['column'].method()`Realiza cada cálculo por separado.

Para evaluar la actividad de los usuarios y las usuarias en cada ciudad, agrupa los datos por ciudad y encuentra la cantidad de canciones reproducidas en cada grupo.



In [34]:
# Contar las canciones reproducidas en cada ciudad


`Comenta tus observaciones aquí`

Ahora agrupemos los datos por día de la semana y encontremos el número de canciones reproducidas el lunes, miércoles y viernes. Utiliza el mismo método que antes, pero ahora necesitamos una agrupación diferente.


In [35]:
# Calcular las canciones reproducidas en cada uno de los tres días


`Comenta tus observaciones aquí`

Ya sabes cómo contar entradas agrupándolas por ciudad o día. Ahora necesitas escribir una función que pueda contar entradas según ambos criterios simultáneamente.

Crea la función `number_tracks()` para calcular el número de canciones reproducidas en un determinado día **y** ciudad. La función debe aceptar dos parámetros:

- `day`: un día de la semana para filtrar. Por ejemplo, `'Monday'` (lunes).
- `city`: una ciudad para filtrar. Por ejemplo, `'Springfield'`.

Dentro de la función, aplicarás un filtrado consecutivo con indexación lógica.

Primero filtra los datos por día y luego filtra la tabla resultante por ciudad.

Después de filtrar los datos por dos criterios, cuenta el número de valores de la columna 'user_id' en la tabla resultante. Este recuento representa el número de entradas que estás buscando. Guarda el resultado en una nueva variable y devuélvelo desde la función.

In [36]:
# Declara la función number_tracks() con dos parámetros: day= y city=.

    # Almacena las filas del DataFrame donde el valor en la columna 'day' es igual al parámetro day=

    # Filtra las filas donde el valor en la columna 'city' es igual al parámetro city=

    # Extrae la columna 'user_id' de la tabla filtrada y aplica el método count()

    # Devolve el número de valores de la columna 'user_id'

Llama a `number_tracks()` seis veces, cambiando los valores de los parámetros para que recuperes los datos de ambas ciudades para cada uno de los tres días.

In [37]:
# El número de canciones reproducidas en Springfield el lunes


In [38]:
# El número de canciones reproducidas en Shelbyville el lunes


In [39]:
# El número de canciones reproducidas en Springfield el miércoles


In [40]:
# El número de canciones reproducidas en Shelbyville el miércoles


In [41]:
# El número de canciones reproducidas en Springfield el viernes


In [42]:
# El número de canciones reproducidas en Shelbyville el viernes


**Conclusiones**

`Comenta si la hipótesis es correcta o se debe rechazar. Explica tu razonamiento.`

[Volver a Contenidos](#back)

# Conclusiones <a id='end'></a>

`Resume aquí tus conclusiones sobre la hipótesis.`

### Nota
En proyectos de investigación reales, la prueba de hipótesis estadística es más precisa y cuantitativa. También ten en cuenta que no siempre se pueden sacar conclusiones sobre una ciudad entera a partir de datos de una sola fuente.

Aprenderás más sobre la prueba de hipótesis en el sprint de análisis estadístico de datos.

[Volver a Contenidos](#back)